src/output.rs

```diff
@@ -353,24 +353,168 @@ impl fmt::Display for DebugBuffer {
     }
 }
 
-fn format_bytes(data: &[u8], f: &mut fmt::Formatter<'_>) -> fmt::Result {
+fn format_bytes(data: &[u8], f: &mut impl fmt::Write) -> fmt::Result {
     #![allow(clippy::assertions_on_constants)]
-    const MIN_OVERFLOW: usize = 8192;
-    const MAX_START: usize = 2048;
-    const MAX_END: usize = 2048;
-    const MAX_PRINTED: usize = MAX_START + MAX_END;
-    assert!(MAX_PRINTED < MIN_OVERFLOW);
 
-    if data.len() >= MIN_OVERFLOW {
+    const LINES_MIN_OVERFLOW: usize = 40;
+    const LINES_MAX_START: usize = 10;
+    const LINES_MAX_END: usize = 10;
+    const LINES_MAX_PRINTED: usize = LINES_MAX_START + LINES_MAX_END;
+    assert!(LINES_MAX_PRINTED < LINES_MIN_OVERFLOW);
+
+    const BYTES_MIN_OVERFLOW: usize = 8192;
+    const BYTES_MAX_START: usize = 2048;
+    const BYTES_MAX_END: usize = 2048;
+    const BYTES_MAX_PRINTED: usize = BYTES_MAX_START + BYTES_MAX_END;
+    assert!(BYTES_MAX_PRINTED < BYTES_MIN_OVERFLOW);
+
+    let data_as_bstr = restore_newlines(&format!("{:?}", data.as_bstr()))?;
+
+    // Strip quotes at beginning and end.
+    let lines = data_as_bstr[1..data_as_bstr.len() - 1]
+        .lines()
+        .collect::<Vec<_>>();
```

Comment:
This takes too many allocations.

1. You can count lines without any allocations and formatting
2. Then you can format each line using bstr while iterating, no need to make a vec
3. Otherwise format like before

Otherwise, logic might be okay. But I'm thinking many `\n` can be in binary data too. So It might be good to ensure that there are at least 10 characters on each line or something to have nicely printed output.

Reply:
> * You can count lines without any allocations and formatting
> * Then you can format each line using bstr while iterating, no need to make a vec

I considered that, but then you throw away the formatting that's already been done.

Is there a compelling reason to prefer reformatting over allocating?

Reply:
Yes, you've got it right. All non-ascii utf-8 chars have have byte codes > 127. So you're safe when scanning for `\n`.

But `.as_bstr()` is also fine. As it's non-allocating actually. It's `format!` that is slow and allocating. So please use `BStr`. It's designed specifically for this sort of thing (of maybe string maybe binary type). And it also has `.lines()` that handles `\r\n` too.

Reply:
@tailhook Thanks for all of the above. I will push an update within the next few days.